File: src/clients/java/src/main/java/com/archerdb/geo/CircuitBreaker.java
package com.archerdb.geo;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Circuit breaker for per-replica failure isolation.
 *
 * <p>
 * Per client-retry/spec.md, the circuit breaker implements:
 * <ul>
 * <li>Three states: Closed (normal), Open (fail-fast), Half-Open (testing recovery)</li>
 * <li>Opens when: 50% failure rate in 10s window AND &gt;= 10 requests</li>
 * <li>Stays open for 30 seconds before transitioning to half-open</li>
 * <li>Half-open allows 5 test requests before deciding to close or re-open</li>
 * <li>Per-replica scope (not global) to allow trying other replicas</li>
 * </ul>
 *
 * <p>
 * This class is thread-safe.
 */
public final class CircuitBreaker {

    /**
     * Circuit breaker states.
     */
    public enum State {
        /**
         * Normal operation - requests are allowed through.
         */
        CLOSED,
        /**
         * Fail-fast mode - requests are rejected immediately.
         */
        OPEN,
        /**
         * Recovery testing - limited requests allowed to test recovery.
         */
        HALF_OPEN
    }

    /**
     * Default failure rate threshold (50%).
     */
    public static final double DEFAULT_FAILURE_THRESHOLD = 0.5;

    /**
     * Default minimum requests in window before circuit can open.
     */
    public static final int DEFAULT_MINIMUM_REQUESTS = 10;

    /**
     * Default sliding window duration in milliseconds (10 seconds).
     */
    public static final long DEFAULT_WINDOW_MS = 10_000;

    /**
     * Default duration to stay open before half-open transition (30 seconds).
     */
    public static final long DEFAULT_OPEN_DURATION_MS = 30_000;

    /**
     * Default number of test requests in half-open state.
     */
    public static final int DEFAULT_HALF_OPEN_REQUESTS = 5;

    private final String name;
    private final double failureThreshold;
    private final int minimumRequests;
    private final long windowMs;
    private final long openDurationMs;
    private final int halfOpenRequests;

    private final AtomicReference<State> state;
    private final AtomicLong openedAt;

    // Sliding window counters
    private final AtomicInteger totalRequests;
    private final AtomicInteger failedRequests;
    private final AtomicLong windowStartMs;

    // Half-open state tracking
    private final AtomicInteger halfOpenSuccesses;
    private final AtomicInteger halfOpenFailures;
    private final AtomicInteger halfOpenTotal;

    // Metrics
    private final AtomicLong stateChanges;
    private final AtomicLong rejectedRequests;

    /**
     * Creates a circuit breaker with default settings.
     *
     * @param name the circuit breaker name (typically replica address)
     */
    public CircuitBreaker(String name) {
        this(new Builder().setName(name));
    }

    private CircuitBreaker(Builder builder) {
        this.name = builder.name;
        this.failureThreshold = builder.failureThreshold;
        this.minimumRequests = builder.minimumRequests;
        this.windowMs = builder.windowMs;
        this.openDurationMs = builder.openDurationMs;
        this.halfOpenRequests = builder.halfOpenRequests;

        this.state = new AtomicReference<>(State.CLOSED);
        this.openedAt = new AtomicLong(0);

        this.totalRequests = new AtomicInteger(0);
        this.failedRequests = new AtomicInteger(0);
        this.windowStartMs = new AtomicLong(System.currentTimeMillis());

        this.halfOpenSuccesses = new AtomicInteger(0);
        this.halfOpenFailures = new AtomicInteger(0);
        this.halfOpenTotal = new AtomicInteger(0);

        this.stateChanges = new AtomicLong(0);
        this.rejectedRequests = new AtomicLong(0);
    }

    /**
     * Checks if a request is allowed through the circuit breaker.
     *
     * @return true if request is allowed, false if circuit is open
     */
    public boolean allowRequest() {
        State currentState = state.get();

        switch (currentState) {
            case CLOSED:
                return true;

            case OPEN:
                // Check if open duration has elapsed
                long elapsed = System.currentTimeMillis() - openedAt.get();
                if (elapsed >= openDurationMs) {
                    // Transition to half-open
                    if (transitionTo(State.HALF_OPEN)) {
                        resetHalfOpenCounters();
                    }
                    return allowHalfOpenRequest();
                }
                rejectedRequests.incrementAndGet();
                return false;

            case HALF_OPEN:
                return allowHalfOpenRequest();

            default:
                return false;
        }
    }

    private boolean allowHalfOpenRequest() {
        int current = halfOpenTotal.get();
        if (current >= halfOpenRequests) {
            // All test slots used - reject
            rejectedRequests.incrementAndGet();
            return false;
        }
        // Allow if we can increment within limit
        return halfOpenTotal.incrementAndGet() <= halfOpenRequests;
    }

    /**
     * Records a successful request.
     */
    public void recordSuccess() {
        State currentState = state.get();

        switch (currentState) {
            case CLOSED:
                recordInWindow(false);
                break;

            case HALF_OPEN:
                int successes = halfOpenSuccesses.incrementAndGet();
                // Check if we've completed all test requests successfully
                if (successes >= halfOpenRequests) {
                    transitionTo(State.CLOSED);
                    resetCounters();
                }
                break;

            case OPEN:
                // Shouldn't happen - request shouldn't have been allowed
                break;
        }
    }

    /**
     * Records a failed request.
     */
    public void recordFailure() {
        State currentState = state.get();

        switch (currentState) {
            case CLOSED:
                recordInWindow(true);
                checkThreshold();
                break;

            case HALF_OPEN:
                halfOpenFailures.incrementAndGet();
                // Any failure in half-open immediately reopens
                transitionTo(State.OPEN);
                break;

            case OPEN:
                // Shouldn't happen - request shouldn't have been allowed
                break;
        }
    }

    private void recordInWindow(boolean failed) {
        long now = System.currentTimeMillis();
        long windowStart = windowStartMs.get();

        // Check if window has expired and reset atomically
        if (now - windowStart >= windowMs && windowStartMs.compareAndSet(windowStart, now)) {
            totalRequests.set(0);
            failedRequests.set(0);
        }

        totalRequests.incrementAndGet();
        if (failed) {
            failedRequests.incrementAndGet();
        }
    }

    private void checkThreshold() {
        int total = totalRequests.get();
        int failed = failedRequests.get();

        // Only check if minimum requests reached
        if (total < minimumRequests) {
            return;
        }

        // Calculate failure rate
        double failureRate = (double) failed / total;

        // Open circuit if threshold exceeded
        if (failureRate >= failureThreshold) {
            transitionTo(State.OPEN);
        }
    }

    private boolean transitionTo(State newState) {
        State current = state.get();
        if (current == newState) {
            return false;
        }

        if (state.compareAndSet(current, newState)) {
            stateChanges.incrementAndGet();

            if (newState == State.OPEN) {
                openedAt.set(System.currentTimeMillis());
            }

            return true;
        }
        return false;
    }

    private void resetCounters() {
        totalRequests.set(0);
        failedRequests.set(0);
        windowStartMs.set(System.currentTimeMillis());
    }

    private void resetHalfOpenCounters() {
        halfOpenSuccesses.set(0);
        halfOpenFailures.set(0);
        halfOpenTotal.set(0);
    }

    /**
     * Returns the current state.
     *
     * @return the state
     */
    public State getState() {
        // Check for automatic transition from OPEN to HALF_OPEN
        if (state.get() == State.OPEN) {
            long elapsed = System.currentTimeMillis() - openedAt.get();
            if (elapsed >= openDurationMs) {
                transitionTo(State.HALF_OPEN);
                resetHalfOpenCounters();
            }
        }
        return state.get();
    }

    /**
     * Returns the circuit breaker name.
     *
     * @return the name
     */
    public String getName() {
        return name;
    }

    /**
     * Returns the total number of state changes.
     *
     * @return state change count
     */
    public long getStateChanges() {
        return stateChanges.get();
    }

    /**
     * Returns the total number of rejected requests.
     *
     * @return rejected request count
     */
    public long getRejectedRequests() {
        return rejectedRequests.get();
    }

    /**
     * Returns the current failure rate in the sliding window.
     *
     * @return failure rate (0.0 to 1.0)
     */
    public double getCurrentFailureRate() {
        int total = totalRequests.get();
        if (total == 0) {
            return 0.0;
        }
        return (double) failedRequests.get() / total;
    }

    /**
     * Forces the circuit breaker to close. Use with caution - primarily for testing.
     */
    public void forceClose() {
        State prev = state.getAndSet(State.CLOSED);
        if (prev != State.CLOSED) {
            stateChanges.incrementAndGet();
        }
        resetCounters();
        resetHalfOpenCounters();
    }

    /**
     * Forces the circuit breaker to open. Use with caution - primarily for testing.
     */
    public void forceOpen() {
        State prev = state.getAndSet(State.OPEN);
        if (prev != State.OPEN) {
            stateChanges.incrementAndGet();
        }
        openedAt.set(System.currentTimeMillis());
    }

    /**
     * Returns true if the circuit is open (rejecting requests).
     *
     * @return true if open
     */
    public boolean isOpen() {
        return getState() == State.OPEN;
    }

    /**
     * Returns true if the circuit is closed (normal operation).
     *
     * @return true if closed
     */
    public boolean isClosed() {
        return getState() == State.CLOSED;
    }

    /**
     * Returns true if the circuit is half-open (testing recovery).
     *
     * @return true if half-open
     */
    public boolean isHalfOpen() {
        return getState() == State.HALF_OPEN;
    }

    @Override
    public String toString() {
        return String.format("CircuitBreaker[name=%s, state=%s, failureRate=%.2f%%]", name,
                getState(), getCurrentFailureRate() * 100);
    }

    /**
     * Creates a new builder.
     *
     * @return the builder
     */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * Builder for CircuitBreaker.
     */
    public static class Builder {
        private String name = "default";
        private double failureThreshold = DEFAULT_FAILURE_THRESHOLD;
        private int minimumRequests = DEFAULT_MINIMUM_REQUESTS;
        private long windowMs = DEFAULT_WINDOW_MS;
        private long openDurationMs = DEFAULT_OPEN_DURATION_MS;
        private int halfOpenRequests = DEFAULT_HALF_OPEN_REQUESTS;

        /**
         * Sets the circuit breaker name.
         *
         * @param name the name (typically replica address)
         * @return this builder
         */
        public Builder setName(String name) {
            if (name == null || name.isEmpty()) {
                throw new IllegalArgumentException("Name cannot be null or empty");
            }
            this.name = name;
            return this;
        }

        /**
         * Sets the failure rate threshold to open the circuit.
         *
         * @param threshold the threshold (0.0 to 1.0)
         * @return this builder
         */
        public Builder setFailureThreshold(double threshold) {
            if (threshold < 0.0 || threshold > 1.0) {
                throw new IllegalArgumentException("Threshold must be between 0.0 and 1.0");
            }
            this.failureThreshold = threshold;
            return this;
        }

        /**
         * Sets the minimum number of requests before the circuit can open.
         *
         * @param minimum the minimum requests
         * @return this builder
         */
        public Builder setMinimumRequests(int minimum) {
            if (minimum <= 0) {
                throw new IllegalArgumentException("Minimum requests must be positive");
            }
            this.minimumRequests = minimum;
            return this;
        }

        /**
         * Sets the sliding window duration.
         *
         * @param windowMs the window in milliseconds
         * @return this builder
         */
        public Builder setWindowMs(long windowMs) {
            if (windowMs <= 0) {
                throw new IllegalArgumentException("Window must be positive");
            }
            this.windowMs = windowMs;
            return this;
        }

        /**
         * Sets how long the circuit stays open before half-open.
         *
         * @param openDurationMs the duration in milliseconds
         * @return this builder
         */
        public Builder setOpenDurationMs(long openDurationMs) {
            if (openDurationMs <= 0) {
                throw new IllegalArgumentException("Open duration must be positive");
            }
            this.openDurationMs = openDurationMs;
            return this;
        }

        /**
         * Sets the number of test requests allowed in half-open state.
         *
         * @param requests the number of test requests
         * @return this builder
         */
        public Builder setHalfOpenRequests(int requests) {
            if (requests <= 0) {
                throw new IllegalArgumentException("Half-open requests must be positive");
            }
            this.halfOpenRequests = requests;
            return this;
        }

        /**
         * Builds the circuit breaker.
         *
         * @return the circuit breaker
         */
        public CircuitBreaker build() {
            return new CircuitBreaker(this);
        }
    }
}

File: src/clients/go/pkg/circuitbreaker/circuitbreaker.go
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2025 ArcherDB Contributors
// Package circuitbreaker provides per-replica circuit breaker for failure isolation.
//
// This package implements the circuit breaker pattern from client-retry/spec.md:
// - Opens when: 50% failure rate in 10s window AND >= 10 requests
// - Stays open for 30 seconds before transitioning to half-open
// - Half-open allows 5 test requests before deciding to close or re-open
// - Per-replica scope (not global) to allow trying other replicas
package circuitbreaker

import (
	"fmt"
	"sync"
	"time"
)

// State represents the circuit breaker state.
type State int

const (
	// Closed allows all requests through, monitoring for failures.
	Closed State = iota
	// Open rejects all requests immediately.
	Open
	// HalfOpen allows limited test requests to probe recovery.
	HalfOpen
)

// String returns the state name.
func (s State) String() string {
	switch s {
	case Closed:
		return "CLOSED"
	case Open:
		return "OPEN"
	case HalfOpen:
		return "HALF_OPEN"
	default:
		return "UNKNOWN"
	}
}

// Config holds circuit breaker configuration options.
type Config struct {
	// FailureThreshold is the failure rate to open circuit (default: 0.5 = 50%).
	FailureThreshold float64

	// MinRequests is the minimum requests in window before circuit can open (default: 10).
	MinRequests int

	// WindowDurationMs is the sliding window duration in milliseconds (default: 10000).
	WindowDurationMs int

	// OpenDurationMs is how long circuit stays open before transitioning to half-open (default: 30000).
	OpenDurationMs int

	// HalfOpenRequests is the number of test requests allowed in half-open state (default: 5).
	HalfOpenRequests int
}

// DefaultConfig returns the default circuit breaker configuration per spec.
func DefaultConfig() Config {
	return Config{
		FailureThreshold: 0.5,
		MinRequests:      10,
		WindowDurationMs: 10000,
		OpenDurationMs:   30000,
		HalfOpenRequests: 5,
	}
}

// ErrCircuitOpen is returned when the circuit breaker is open.
type ErrCircuitOpen struct {
	CircuitName  string
	CircuitState State
}

func (e ErrCircuitOpen) Error() string {
	return fmt.Sprintf("circuit breaker '%s' is %s - request rejected", e.CircuitName, e.CircuitState)
}

// CircuitBreaker provides per-replica circuit breaker for failure isolation.
//
// Per client-retry/spec.md:
// - Opens when: 50% failure rate in 10s window AND >= 10 requests
// - Stays open for 30 seconds before transitioning to half-open
// - Half-open allows 5 test requests before deciding to close or re-open
// - Per-replica scope (not global) to allow trying other replicas
//
// Thread-safe implementation.
type CircuitBreaker struct {
	name   string
	config Config

	mu       sync.Mutex
	state    State
	openedAt time.Time

	// Sliding window counters
	totalRequests   int
	failedRequests  int
	windowStartTime time.Time

	// Half-open tracking
	halfOpenSuccesses int
	halfOpenFailures  int
	halfOpenTotal     int

	// Metrics
	stateChanges     int
	rejectedRequests int
}

// New creates a new CircuitBreaker with the given name and default config.
func New(name string) *CircuitBreaker {
	return NewWithConfig(name, DefaultConfig())
}

// NewWithConfig creates a new CircuitBreaker with custom configuration.
func NewWithConfig(name string, config Config) *CircuitBreaker {
	return &CircuitBreaker{
		name:            name,
		config:          config,
		state:           Closed,
		windowStartTime: time.Now(),
	}
}

// Name returns the circuit breaker name.
func (cb *CircuitBreaker) Name() string {
	return cb.name
}

// State returns the current circuit state, checking for automatic transitions.
func (cb *CircuitBreaker) State() State {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	if cb.state == Open {
		elapsed := time.Since(cb.openedAt)
		if elapsed >= time.Duration(cb.config.OpenDurationMs)*time.Millisecond {
			cb.transitionTo(HalfOpen)
			cb.resetHalfOpenCounters()
		}
	}
	return cb.state
}

// AllowRequest checks if a request is allowed through.
func (cb *CircuitBreaker) AllowRequest() bool {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	switch cb.state {
	case Closed:
		return true

	case Open:
		elapsed := time.Since(cb.openedAt)
		if elapsed >= time.Duration(cb.config.OpenDurationMs)*time.Millisecond {
			cb.transitionTo(HalfOpen)
			cb.resetHalfOpenCounters()
			return cb.allowHalfOpenRequest()
		}
		cb.rejectedRequests++
		return false

	case HalfOpen:
		return cb.allowHalfOpenRequest()
	}

	return false
}

func (cb *CircuitBreaker) allowHalfOpenRequest() bool {
	if cb.halfOpenTotal >= cb.config.HalfOpenRequests {
		cb.rejectedRequests++
		return false
	}
	cb.halfOpenTotal++
	return true
}

// RecordSuccess records a successful request.
func (cb *CircuitBreaker) RecordSuccess() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	switch cb.state {
	case Closed:
		cb.recordInWindow(false)

	case HalfOpen:
		cb.halfOpenSuccesses++
		if cb.halfOpenSuccesses >= cb.config.HalfOpenRequests {
			cb.transitionTo(Closed)
			cb.resetCounters()
		}
	}
}

// RecordFailure records a failed request.
func (cb *CircuitBreaker) RecordFailure() {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	switch cb.state {
	case Closed:
		cb.recordInWindow(true)
		cb.checkThreshold()

	case HalfOpen:
		cb.halfOpenFailures++
		cb.transitionTo(Open)
	}
}

func (cb *CircuitBreaker) recordInWindow(failed bool) {
	now := time.Now()
	elapsed := now.Sub(cb.windowStartTime)

	// Reset window if expired
	if elapsed >= time.Duration(cb.config.WindowDurationMs)*time.Millisecond {
		cb.totalRequests = 0
		cb.failedRequests = 0
		cb.windowStartTime = now
	}

	cb.totalRequests++
	if failed {
		cb.failedRequests++
	}
}

func (cb *CircuitBreaker) checkThreshold() {
	if cb.totalRequests < cb.config.MinRequests {
		return
	}

	failureRate := float64(cb.failedRequests) / float64(cb.totalRequests)
	if failureRate >= cb.config.FailureThreshold {
		cb.transitionTo(Open)
	}
}

func (cb *CircuitBreaker) transitionTo(newState State) {
	if cb.state == newState {
		return
	}
	cb.state = newState
	cb.stateChanges++

	if newState == Open {
		cb.openedAt = time.Now()
	}
}

func (cb *CircuitBreaker) resetCounters() {
	cb.totalRequests = 0
	cb.failedRequests = 0
	cb.windowStartTime = time.Now()
}

func (cb *CircuitBreaker) resetHalfOpenCounters() {
	cb.halfOpenSuccesses = 0
	cb.halfOpenFailures = 0
	cb.halfOpenTotal = 0
}

// FailureRate returns the current failure rate in the sliding window.
func (cb *CircuitBreaker) FailureRate() float64 {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	if cb.totalRequests == 0 {
		return 0.0
	}
	return float64(cb.failedRequests) / float64(cb.totalRequests)
}

// Metrics returns circuit breaker metrics.
func (cb *CircuitBreaker) Metrics() Metrics {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	failureRate := 0.0
	if cb.totalRequests > 0 {
		failureRate = float64(cb.failedRequests) / float64(cb.totalRequests)
	}

	return Metrics{
		State:            cb.state,
		TotalRequests:    cb.totalRequests,
		FailedRequests:   cb.failedRequests,
		StateChanges:     cb.stateChanges,
		RejectedRequests: cb.rejectedRequests,
		FailureRate:      failureRate,
	}
}

// Metrics holds circuit breaker metrics.
type Metrics struct {
	State            State
	TotalRequests    int
	FailedRequests   int
	StateChanges     int
	RejectedRequests int
	FailureRate      float64
}

// ForceOpen forces the circuit open (for testing).
func (cb *CircuitBreaker) ForceOpen() {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.transitionTo(Open)
}

// ForceClosed forces the circuit closed (for testing).
func (cb *CircuitBreaker) ForceClosed() {
	cb.mu.Lock()
	defer cb.mu.Unlock()
	cb.transitionTo(Closed)
	cb.resetCounters()
}

// IsOpen returns true if the circuit is open.
func (cb *CircuitBreaker) IsOpen() bool {
	return cb.State() == Open
}

// IsClosed returns true if the circuit is closed.
func (cb *CircuitBreaker) IsClosed() bool {
	return cb.State() == Closed
}

// IsHalfOpen returns true if the circuit is half-open.
func (cb *CircuitBreaker) IsHalfOpen() bool {
	return cb.State() == HalfOpen
}

func (cb *CircuitBreaker) String() string {
	cb.mu.Lock()
	defer cb.mu.Unlock()

	failureRate := 0.0
	if cb.totalRequests > 0 {
		failureRate = float64(cb.failedRequests) / float64(cb.totalRequests) * 100
	}

	return fmt.Sprintf("CircuitBreaker[name=%s, state=%s, failureRate=%.2f%%]",
		cb.name, cb.state, failureRate)
}
